storage/badger/protocol_kv_store.go

```diff
@@ -0,0 +1,164 @@
+			withLimit[flow.Identifier, *storage.KeyValueStoreData](kvStoreCacheSize),
+			withStore(storeByStateID),
+			withRetrieve(retrieveByStateID)),
+		byBlockIdCache: newCache[flow.Identifier, flow.Identifier](collector, metrics.ResourceProtocolKVStoreByBlockID,
+			withLimit[flow.Identifier, flow.Identifier](kvStoreByBlockIDCacheSize),
+			withStore(storeByBlockID),
+			withRetrieve(retrieveByBlockID)),
+	}
+}
+
+// StoreTx returns an anonymous function (intended to be executed as part of a badger transaction),
+// which persists the given model as part of a DB tx.
+// Expected errors of the returned anonymous function:
+//   - storage.ErrAlreadyExists if a model with the given id is already stored
+func (s *ProtocolKVStore) StoreTx(stateID flow.Identifier, data *storage.KeyValueStoreData) func(*transaction.Tx) error {
+	return s.cache.PutTx(stateID, data)
+}
+
+// IndexTx returns an anonymous function intended to be executed as part of a database transaction.
+// In a nutshell, we want to maintain a map from `blockID` to `stateID`, where `blockID` references the
+// block that _proposes_ updated key-value store.
+// Upon call, the anonymous function persists the specific map entry in the node's database.
+// Protocol convention:
+//   - Consider block B, whose ingestion might potentially lead to an updated KV store. For example,
+//     the KV store changes if we seal some execution results emitting specific service events.
+//   - For the key `blockID`, we use the identity of block B which _proposes_ this updated KV store. As value,
+//     the hash of the resulting key-value store state at the end of processing B is to be used.
+//   - CAUTION: The updated state requires confirmation by a QC and will only become active at the child block,
+//     _after_ validating the QC.
+//
+// Expected errors during normal operations:
+//   - storage.ErrAlreadyExists if a KV store for the given blockID has already been indexed
+func (s *ProtocolKVStore) IndexTx(blockID flow.Identifier, stateID flow.Identifier) func(*transaction.Tx) error {
+	return s.byBlockIdCache.PutTx(blockID, stateID)
+}
+
+// ByID returns the key-value store model by its ID.
+// Expected errors during normal operations:
+//   - storage.ErrNotFound if no model with the given Identifier is known.
+func (s *ProtocolKVStore) ByID(id flow.Identifier) (*storage.KeyValueStoreData, error) {
+	tx := s.db.NewTransaction(false)
+	defer tx.Discard()
+	return s.cache.Get(id)(tx)
+}
+
+// ByBlockID retrieves the key-value store model that the block with the given ID proposes.
+// CAUTION: this store state requires confirmation by a QC and will only become active at the child block,
+// _after_ validating the QC. Protocol convention:
+//   - Consider block B, whose ingestion might potentially lead to an updated KV store state.
+//
+// For example, the state changes if we seal some execution results emitting specific service events.
```

Comment:
I would indent this, so the bullet point reads:
> * Consider block B, whose ingestion might potentially lead to an updated KV store state. For example, the state changes if we seal some execution results emitting specific service events.

```suggestion
//   - Consider block B, whose ingestion might potentially lead to an updated KV store state.
//     For example, the state changes if we seal some execution results emitting specific service events.
```